File: onnxruntime_cpp/oop_anomaly_detector.cpp
#include "oop_anomaly_detector.h"
#include <iostream>
#include <algorithm>
#include <chrono>
#include <sstream>

// -------------------- Config Implementation --------------------
bool Config::parse(int argc, char** argv) {
    if(argc < 3) return false;
    model = argv[1];
    images_dir = argv[2];
    for(int i=3;i<argc;i++){
        std::string a = argv[i];
        if(a=="--save_viz" && i+1<argc){ save_viz=true; out_dir=argv[++i]; }
        else if(a=="--alpha" && i+1<argc) alpha = std::stod(argv[++i]);
        else if(a=="--thresh" && i+1<argc) thresh = std::stof(argv[++i]);
    }
    if(!fs::exists(model) || !fs::exists(images_dir)) return false;
    if(save_viz && !fs::exists(out_dir)) fs::create_directories(out_dir);
    return true;
}

// -------------------- ImageList Implementation --------------------
std::vector<fs::path> ImageList::list(const fs::path& dir) {
    std::vector<fs::path> imgs;
    std::vector<std::string> exts{".jpg",".jpeg",".png",".bmp",".tiff"};
    if(!fs::exists(dir) || !fs::is_directory(dir)) return imgs;
    for(auto &e: fs::directory_iterator(dir)){
        if(!fs::is_regular_file(e.path())) continue;
        std::string ext = e.path().extension().string();
        std::transform(ext.begin(),ext.end(),ext.begin(),::tolower);
        if(std::find(exts.begin(),exts.end(),ext)!=exts.end()) imgs.push_back(e.path());
    }
    std::sort(imgs.begin(), imgs.end());
    return imgs;
}

// -------------------- Preprocessor Implementation --------------------
Preprocessor::Preprocessor(int H, int W, const std::array<float,3>& mean, const std::array<float,3>& stdev)
    : H_(H), W_(W), mean_(mean), stdev_(stdev) {}

std::vector<float> Preprocessor::preprocess(const cv::Mat& bgr) const {
    cv::Mat rgb; cv::cvtColor(bgr, rgb, cv::COLOR_BGR2RGB);
    cv::Mat r; cv::resize(rgb, r, cv::Size(W_,H_));
    r.convertTo(r, CV_32FC3, 1.0f/255.0f);
    std::vector<cv::Mat> ch(3); cv::split(r, ch);
    std::vector<float> out; out.reserve(3*H_*W_);
    for(int c=0;c<3;++c){
        const cv::Mat& m = ch[c];
        for(int y=0;y<H_;++y){
            const float* ptr = m.ptr<float>(y);
            for(int x=0;x<W_;++x) out.push_back((ptr[x]-mean_[c])/stdev_[c]);
        }
    }
    return out;
}

// -------------------- ONNXModel Implementation --------------------
ONNXModel::ONNXModel(const std::string& model_path, Ort::Env& env, Ort::SessionOptions& opts)
    : env_(env), session_(nullptr), alloc_() {
#ifdef _WIN32
    std::wstring wmodel(model_path.begin(), model_path.end());
    session_ = std::make_unique<Ort::Session>(env_, wmodel.c_str(), opts);
#else
    session_ = std::make_unique<Ort::Session>(env_, model_path.c_str(), opts);
#endif
    fetch_io();
}

const std::string& ONNXModel::input_name() const { return input_name_; }
const std::vector<int64_t>& ONNXModel::input_shape() const { return input_shape_; }
const std::vector<std::string>& ONNXModel::output_names() const { return output_names_; }

std::vector<Ort::Value> ONNXModel::run(float* input_data, size_t input_count) {
    std::vector<int64_t> dims = input_shape_;
    if(dims.size()>0 && dims[0] <= 0) dims[0] = 1;

    Ort::MemoryInfo meminfo = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
    Ort::Value input_tensor = Ort::Value::CreateTensor<float>(meminfo, input_data, input_count, dims.data(), dims.size());

    std::vector<const char*> in_names = { input_name_.c_str() };
    std::vector<const char*> out_names_c;
    out_names_c.reserve(output_names_.size());
    for(const auto &n: output_names_) out_names_c.push_back(n.c_str());

    auto outputs = session_->Run(Ort::RunOptions{nullptr}, in_names.data(), &input_tensor, 1, out_names_c.data(), out_names_c.size());
    return outputs;
}

void ONNXModel::fetch_io() {
    char* in = session_->GetInputName(0, alloc_);
    input_name_ = in;
    alloc_.Free(in);

    auto info = session_->GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape();
    input_shape_.assign(info.begin(), info.end());

    size_t nout = session_->GetOutputCount();
    for(size_t i=0;i<nout;i++){
        char* n = session_->GetOutputName(i, alloc_);
        output_names_.emplace_back(n);
        alloc_.Free(n);
    }
}

// -------------------- Postprocessor Implementation --------------------
Postprocessor::Postprocessor(float threshold) : threshold_(threshold) {}

void Postprocessor::process(std::vector<Ort::Value>& outputs, Result& r) {
    if(outputs.size() >= 2){
        float* sdata = outputs[0].GetTensorMutableData<float>();
        r.score = sdata[0];
        auto shape = outputs[1].GetTensorTypeAndShapeInfo().GetShape();
        float* hdata = outputs[1].GetTensorMutableData<float>();
        int H = (int)shape[shape.size()-2], W = (int)shape[shape.size()-1];
        r.heatmap = cv::Mat(H,W,CV_32FC1,hdata).clone();
    } else if(outputs.size() == 1){
        auto &t = outputs[0];
        auto shape = t.GetTensorTypeAndShapeInfo().GetShape();
        int64_t total=1; for(auto d:shape) if(d>0) total*=d;
        float* data = t.GetTensorMutableData<float>();
        if(total==1) r.score = data[0];
        else {
            int H = (int)shape[shape.size()-2], W = (int)shape[shape.size()-1];
            r.heatmap = cv::Mat(H,W,CV_32FC1,data).clone();
            cv::Scalar m = cv::mean(r.heatmap); r.score = (float)m[0];
        }
    }

    if(!r.heatmap.empty()){
        cv::GaussianBlur(r.heatmap, r.heatmap, cv::Size(33,33), 4.0);
        cv::threshold(r.heatmap, r.mask, threshold_, 255.0, cv::THRESH_BINARY);
        r.mask.convertTo(r.mask, CV_8UC1);
        r.pixels = cv::countNonZero(r.mask);
        r.ratio = (float)r.pixels / (r.heatmap.rows * r.heatmap.cols);
    }
    r.anomalous = r.score > threshold_;
}

// -------------------- Visualizer Implementation --------------------
Visualizer::Visualizer(double alpha) : alpha_(alpha) {}

cv::Mat Visualizer::overlay(const cv::Mat& orig, const cv::Mat& heat) const {
    // Just return original image (no heatmap overlay)
    return orig.clone();
}


void Visualizer::annotate(cv::Mat& img, const Result& r, double t_ms) const {
    cv::Scalar txt = r.anomalous ? cv::Scalar(0,0,255) : cv::Scalar(0,255,0);
    cv::rectangle(img, {5,5}, {420,110}, {0,0,0}, -1);
    cv::putText(img, "Score: "+to_fixed(r.score,3), {10,30}, cv::FONT_HERSHEY_SIMPLEX, 0.7, txt, 2);
    cv::putText(img, r.anomalous?"ANOMALOUS":"NORMAL", {10,60}, cv::FONT_HERSHEY_SIMPLEX, 0.9, txt, 2);
    cv::putText(img, "Time: "+to_fixed(t_ms,2)+" ms", {10,90}, cv::FONT_HERSHEY_SIMPLEX, 0.6, {255,255,255}, 2);
}

void Visualizer::draw_boundaries(cv::Mat& img, const cv::Mat& mask) const {
    if(mask.empty()) return;
    cv::Mat rm; cv::resize(mask, rm, img.size());
    std::vector<std::vector<cv::Point>> ctrs;
    cv::findContours(rm, ctrs, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
    for(size_t i=0;i<ctrs.size();++i){
        if(cv::contourArea(ctrs[i])>50){
            cv::drawContours(img, ctrs, (int)i, {0,255,0}, 3);
            cv::Rect b = cv::boundingRect(ctrs[i]); cv::rectangle(img, b, {0,0,255}, 2);
        }
    }
}

std::string Visualizer::to_fixed(double v, int prec) {
    std::ostringstream oss;
    oss.setf(std::ios::fixed); oss.precision(prec);
    oss<<v;
    return oss.str();
}

// -------------------- App Implementation --------------------
App::App(const Config& cfg) : cfg_(cfg), env_(ORT_LOGGING_LEVEL_WARNING, "det"), opts_(), alloc_() {
    opts_.SetIntraOpNumThreads(1);
    opts_.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);

    model_ = std::make_unique<ONNXModel>(cfg_.model, env_, opts_);

    auto ishape = model_->input_shape();
    if(ishape.size()<4) throw std::runtime_error("Unexpected input shape from model");
    H_ = (int)ishape[2]; W_ = (int)ishape[3];

    preprocessor_ = std::make_unique<Preprocessor>(H_, W_, cfg_.mean, cfg_.std);
    postprocessor_ = std::make_unique<Postprocessor>(cfg_.thresh);
    visualizer_ = std::make_unique<Visualizer>(cfg_.alpha);
}

int App::run() {
    auto images = ImageList::list(cfg_.images_dir);
    if(images.empty()){ std::cerr<<"No images found\n"; return 1; }
    std::cout<<"Found "<<images.size()<<" images\n";

    std::vector<int64_t> dims = model_->input_shape();
    if(dims.size()>0 && dims[0] <= 0) dims[0] = 1;
    size_t input_count = 1; for(auto d: dims) input_count *= (d>0? d: 1);

    for(size_t i=0;i<images.size();++i){
        auto p = images[i];
        std::cout<<"\nProcessing "<<p.filename().string()<<" ("<<i+1<<"/"<<images.size()<<")\n";
        cv::Mat img = cv::imread(p.string(), cv::IMREAD_COLOR);
        if(img.empty()){ std::cerr<<"Failed to open "<<p<<"\n"; continue; }

        auto input_vec = preprocessor_->preprocess(img);
        if(input_vec.size() != input_count){
            if(input_vec.size() > input_count){
                std::cerr<<"Warning: produced more input elements than model expects. Using first N elements.\n";
            } else {
                std::cerr<<"Warning: input size mismatch: produced "<<input_vec.size()<<" expected "<<input_count<<"\n";
            }
        }

        auto t1 = std::chrono::high_resolution_clock::now();
        auto outputs = model_->run(input_vec.data(), std::min(input_vec.size(), input_count));
        auto t2 = std::chrono::high_resolution_clock::now();
        double elapsed_ms = std::chrono::duration_cast<std::chrono::microseconds>(t2-t1).count()/1000.0;

        Result res; res.time_ms = elapsed_ms;
        postprocessor_->process(outputs, res);

        cv::Mat disp = visualizer_->overlay(img, res.heatmap);
        visualizer_->annotate(disp, res, elapsed_ms);
        visualizer_->draw_boundaries(disp, res.mask);

        std::string wname = "Result: " + p.filename().string();
        cv::namedWindow(wname, cv::WINDOW_NORMAL);
        cv::resizeWindow(wname, 800, 600);
        cv::imshow(wname, disp);
        if(cfg_.save_viz){
            fs::path outp = cfg_.out_dir / (p.stem().string() + "_res.png");
            cv::imwrite(outp.string(), disp);
        }
        int key = cv::waitKey(1);
        cv::destroyWindow(wname);
        if(key==27 || key=='q' || key=='Q'){ std::cout<<"User quit\n"; break; }

        std::cout<<"Score: "<<res.score<<"  Anomalous: "<<(res.anomalous?"YES":"NO")<<"  Time: "<<res.time_ms<<" ms\n";
        std::cout<<"Number of anomalous pixels: "<<res.pixels<<"  Ratio to total pixels: "<<res.ratio<<" ms\n";
    }

    std::cout<<"\nDone\n";
    return 0;
}

File: onnxruntime_cpp/oop_anomaly_detector.py
# oop_anomaly_detector.py
# Requires: pip install onnxruntime opencv-python numpy

import argparse
import glob
import os
import sys
import time
from dataclasses import dataclass
from pathlib import Path

import cv2
import numpy as np
import onnxruntime as ort


# -------------------- Config --------------------
@dataclass
class Config:
    model: str
    images_dir: Path
    out_dir: Path = Path("out")
    save_viz: bool = False
    alpha: float = 0.5
    thresh: float = 13.0
    mean: tuple = (0.485, 0.456, 0.406)
    std: tuple = (0.229, 0.224, 0.225)

    @staticmethod
    def parse():
        p = argparse.ArgumentParser(
            description="Python OOP Anomaly Detector (vis compatible with C++)."
        )
        p.add_argument("--model", type=str, help="Path to ONNX model")
        p.add_argument("--images_dir", type=Path, help="Directory of input images")
        p.add_argument("--save_viz", type=Path, help="Directory to save visualizations")
        p.add_argument(
            "--alpha",
            type=float,
            default=0.5,
            help="Overlay alpha (0 disables heat overlay)",
        )
        p.add_argument("--thresh", type=float, default=13.0, help="Anomaly threshold")
        args = p.parse_args()

        if not Path(args.model).exists():
            sys.exit(f"Model not found: {args.model}")
        if not args.images_dir.exists():
            sys.exit(f"Images dir not found: {args.images_dir}")

        save = args.save_viz is not None
        out_dir = args.save_viz if save else Path("out")
        if save:
            out_dir.mkdir(parents=True, exist_ok=True)

        return Config(
            model=args.model,
            images_dir=args.images_dir,
            out_dir=out_dir,
            save_viz=save,
            alpha=args.alpha,
            thresh=float(args.thresh),
        )


# -------------------- Result --------------------
@dataclass
class Result:
    score: float = 0.0
    heatmap: np.ndarray | None = None
    mask: np.ndarray | None = None
    anomalous: bool = False
    pixels: int = 0
    ratio: float = 0.0
    time_ms: float = 0.0


# -------------------- ImageList --------------------
class ImageList:
    exts = {".jpg", ".jpeg", ".png", ".bmp", ".tiff", ".tif"}

    @staticmethod
    def list(dirpath: Path):
        files = []
        for ext in ImageList.exts:
            files.extend(sorted(map(Path, glob.glob(str(dirpath / f"*{ext}")))))
        return files


# -------------------- Preprocessor --------------------
class Preprocessor:
    def __init__(self, H: int, W: int, mean: tuple, stdev: tuple):
        self.H, self.W = H, W
        self.mean = np.array(mean, dtype=np.float32).reshape(3, 1, 1)
        self.std = np.array(stdev, dtype=np.float32).reshape(3, 1, 1)

    def preprocess(self, bgr: np.ndarray) -> np.ndarray:
        rgb = cv2.cvtColor(bgr, cv2.COLOR_BGR2RGB)
        r = cv2.resize(rgb, (self.W, self.H), interpolation=cv2.INTER_LINEAR)
        r = r.astype(np.float32) / 255.0
        r = r.transpose(2, 0, 1)  # HWC->CHW
        r = (r - self.mean) / self.std
        return r  # CHW float32


# -------------------- ONNXModel --------------------
class ONNXModel:
    def __init__(self, model_path: str):
        so = ort.SessionOptions()
        so.intra_op_num_threads = 1
        so.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
        self.session = ort.InferenceSession(
            model_path, sess_options=so, providers=["CPUExecutionProvider"]
        )
        self._input_name = self.session.get_inputs()[0].name
        self._input_shape = self.session.get_inputs()[
            0
        ].shape  # e.g. [1,3,H,W] or [N,3,H,W]
        self._output_names = [o.name for o in self.session.get_outputs()]

    @property
    def input_name(self):
        return self._input_name

    @property
    def input_shape(self):
        return self._input_shape

    @property
    def output_names(self):
        return self._output_names

    def run(self, input_array: np.ndarray):
        # Ensure batch dimension = 1 if dynamic/None
        shape = list(self._input_shape)
        if shape[0] in (None, "None", -1):
            shape[0] = 1
        # Expect CHW -> add batch
        x = np.expand_dims(input_array, 0).astype(np.float32)  # [1,3,H,W]
        feeds = {self._input_name: x}
        return self.session.run(self._output_names, feeds)


# -------------------- Postprocessor --------------------
class Postprocessor:
    def __init__(self, threshold: float):
        self.threshold = float(threshold)

    def process(self, outputs: list) -> Result:
        r = Result()
        # C++ logic: either (score, heatmap) or single tensor that is scalar or HxW
        if len(outputs) >= 2:
            score = float(np.array(outputs[0]).reshape(-1)[0])
            heat = np.array(outputs[1]).squeeze()
            r.score = score
            if heat.ndim == 3:  # e.g. 1xHxW
                heat = heat[0]
            r.heatmap = heat.astype(np.float32)
        elif len(outputs) == 1:
            t = np.array(outputs[0]).squeeze()
            if t.size == 1:
                r.score = float(t)
            else:
                # assume heatmap; score = mean(heatmap)
                if t.ndim == 3:
                    t = t[0]
                r.heatmap = t.astype(np.float32)
                r.score = float(r.heatmap.mean())
        # Extra: blur + threshold + stats (match C++)
        if r.heatmap is not None:
            r.heatmap = cv2.GaussianBlur(r.heatmap, (33, 33), 4.0)
            _, mask = cv2.threshold(r.heatmap, self.threshold, 255.0, cv2.THRESH_BINARY)
            r.mask = mask.astype(np.uint8)
            r.pixels = int(cv2.countNonZero(r.mask))
            h, w = r.heatmap.shape[-2:]
            r.ratio = r.pixels / float(h * w)
        r.anomalous = r.score > self.threshold
        return r


# -------------------- Visualizer --------------------
class Visualizer:
    def __init__(self, alpha: float):
        self.alpha = float(alpha)

    def overlay(self, orig: np.ndarray, heat: np.ndarray | None) -> np.ndarray:

        return orig

    @staticmethod
    def _to_fixed(v: float, prec: int) -> str:
        return f"{v:.{prec}f}"

    def annotate(self, img: np.ndarray, r: Result, t_ms: float):
        txt_color = (0, 0, 255) if r.anomalous else (0, 255, 0)
        cv2.rectangle(img, (5, 5), (420, 110), (0, 0, 0), -1)
        cv2.putText(
            img,
            f"Score: {self._to_fixed(r.score,3)}",
            (10, 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.7,
            txt_color,
            2,
        )
        cv2.putText(
            img,
            "ANOMALOUS" if r.anomalous else "NORMAL",
            (10, 60),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.9,
            txt_color,
            2,
        )
        cv2.putText(
            img,
            f"Time: {self._to_fixed(t_ms,2)} ms",
            (10, 90),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255),
            2,
        )

    def draw_boundaries(self, img: np.ndarray, mask: np.ndarray | None):
        if mask is None:
            return
        rm = cv2.resize(
            mask, (img.shape[1], img.shape[0]), interpolation=cv2.INTER_NEAREST
        )
        contours, _ = cv2.findContours(rm, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        for c in contours:
            if cv2.contourArea(c) > 50:
                cv2.drawContours(img, [c], -1, (0, 255, 0), 3)
                x, y, w, h = cv2.boundingRect(c)
                cv2.rectangle(img, (x, y), (x + w, y + h), (0, 0, 255), 2)


# -------------------- App --------------------
class App:
    def __init__(self, cfg: Config):
        self.cfg = cfg
        self.model = ONNXModel(cfg.model)

        # Expect input [N,3,H,W]
        ishape = self.model.input_shape
        if len(ishape) < 4:
            raise RuntimeError(f"Unexpected input shape: {ishape}")
        self.H = int(ishape[2] if ishape[2] not in (None, -1) else 224)
        self.W = int(ishape[3] if ishape[3] not in (None, -1) else 224)

        self.pre = Preprocessor(self.H, self.W, cfg.mean, cfg.std)
        self.post = Postprocessor(cfg.thresh)
        self.vis = Visualizer(cfg.alpha)

    def run(self):
        images = ImageList.list(self.cfg.images_dir)
        if not images:
            print("No images found")
            return 1
        print(f"Found {len(images)} images")

        for i, p in enumerate(images, 1):
            print(f"\nProcessing {p.name} ({i}/{len(images)})")
            img = cv2.imread(str(p), cv2.IMREAD_COLOR)
            if img is None:
                print(f"Failed to open {p}")
                continue

            x = self.pre.preprocess(img)

            t0 = time.perf_counter()
            outputs = self.model.run(x)
            t1 = time.perf_counter()
            elapsed_ms = (t1 - t0) * 1000.0

            res = self.post.process(outputs)
            res.time_ms = elapsed_ms

            disp = self.vis.overlay(img, res.heatmap)
            self.vis.annotate(disp, res, elapsed_ms)
            self.vis.draw_boundaries(disp, res.mask)

            win = f"Result: {p.name}"
            cv2.namedWindow(win, cv2.WINDOW_NORMAL)
            cv2.resizeWindow(win, 800, 600)
            cv2.imshow(win, disp)

            if self.cfg.save_viz:
                outp = self.cfg.out_dir / f"{p.stem}_res.png"
                cv2.imwrite(str(outp), disp)

            key = cv2.waitKey(1) & 0xFF
            cv2.destroyWindow(win)
            if key in (27, ord("q"), ord("Q")):
                print("User quit")
                break

            print(
                f"Score: {res.score:.4f}  Anomalous: {'YES' if res.anomalous else 'NO'}  Time: {res.time_ms:.2f} ms"
            )
            print(
                f"Number of anomalous pixels: {res.pixels}  Ratio to total pixels: {res.ratio:.6f}"
            )

        print("\nDone")
        return 0


# -------------------- Main --------------------
if __name__ == "__main__":
    cfg = Config.parse()
    sys.exit(App(cfg).run())
# python .\oop_anomaly_detector.py  model_int8.onnx D:/01-DATA/test --save_viz D:/output --alpha 0.5 --thresh 13.0
